TxByPriceAndNonce comparison is logically incorrect

TxByPriceAndNonce : core/types/transaction.go

Say, I have three transactions in the queue with the same price:
A: from account x, nonce=1
B: from account x, nonce=2
C: from account y, nonce=3

according to the comparison function:
A<B = true (comparing by nonce)
B<C = false (comparing by price)
A<C = false (comparing by price)

If the purpose if to "order by price, nonce", then the code should be changed accordingly. Otherwise the result of sort is unpredictable.

```
func (s TxByPriceAndNonce) Less(i, j int) bool {
	comparedByPrice := s.Transactions[i].data.Price.Cmp(s.Transactions[j].data.Price)
	
	if comparedByPrice == 0 {
		return s.Transactions[i].data.AccountNonce < s.Transactions[j].data.AccountNonce
	}
	
	return comparedByPrice > 0
}
```

Here's a test to repro the error and catch all the inconsistencies that should not be there. Quicksort pretty much screws up the order spectacularly :D I think the reason we didn't catch this is because irrelevant of the ordering, there are always some transactions that can actually go through, which will change the order in the next iteration; so after a while, all will manage to pass. Also this issue only surfaces when there are many transactions from many accounts, which is quite rare currently.

```go
// Tests that transactions can be correctly sorted according to their price in
// decreasing order, but at the same time with increasing nonces when issued by
// the same account.
func TestTransactionPriceNonceSort(t *testing.T) {
	// Generate a batch of accounts to start with
	keys := make([]*ecdsa.PrivateKey, 5)
	for i := 0; i < len(keys); i++ {
		keys[i], _ = crypto.GenerateKey()
	}
	// Generate a batch of transactions with overlapping values, but shifted nonces
	txs := []*Transaction{}
	for start, key := range keys {
		for i := 0; i < 5; i++ {
			tx, _ := NewTransaction(uint64(start+i), common.Address{}, big.NewInt(100), big.NewInt(100), big.NewInt(int64(start+i)), nil).SignECDSA(key)
			txs = append(txs, tx)
		}
	}
	// Sort the transactions and cross check the nonce ordering
	sort.Sort(TxByPriceAndNonce{txs})
	for i, txi := range txs {
		fromi, _ := txi.From()

		// Make sure the nonce order is valid
		for j, txj := range txs[i+1:] {
			fromj, _ := txj.From()

			if fromi == fromj && txi.Nonce() > txj.Nonce() {
				t.Errorf("invalid nonce ordering: tx #%d (A=%x N=%v) < tx #%d (A=%x N=%v)", i, fromi[:4], txi.Nonce(), i+j, fromj[:4], txj.Nonce())
			}
		}
		// Find the previous and next nonce of this account
		prev, next := i-1, i+1
		for j := i - 1; j >= 0; j-- {
			if fromj, _ := txs[j].From(); fromi == fromj {
				prev = j
				break
			}
		}
		for j := i + 1; j < len(txs); j++ {
			if fromj, _ := txs[j].From(); fromi == fromj {
				next = j
				break
			}
		}
		// Make sure that in between the neighbor nonces, the transaction is correctly positioned price wise
		for j := prev + 1; j < next; j++ {
			fromj, _ := txs[j].From()
			if j < i && txs[j].GasPrice().Cmp(txi.GasPrice()) < 0 {
				t.Errorf("invalid gasprice ordering: tx #%d (A=%x P=%v) < tx #%d (A=%x P=%v)", j, fromj[:4], txs[j].GasPrice(), i, fromi[:4], txi.GasPrice())
			}
			if j > i && txs[j].GasPrice().Cmp(txi.GasPrice()) > 0 {
				t.Errorf("invalid gasprice ordering: tx #%d (A=%x P=%v) > tx #%d (A=%x P=%v)", j, fromj[:4], txs[j].GasPrice(), i, fromi[:4], txi.GasPrice())
			}
		}
	}
}
```

```
--- FAIL: TestTransactionPriceNonceSort (0.01s)
	transaction_test.go:149: invalid nonce ordering: tx #0 (A=64ba7325 N=8) < tx #2 (A=64ba7325 N=6)
	transaction_test.go:149: invalid nonce ordering: tx #0 (A=64ba7325 N=8) < tx #3 (A=64ba7325 N=7)
	transaction_test.go:149: invalid nonce ordering: tx #0 (A=64ba7325 N=8) < tx #5 (A=64ba7325 N=5)
	transaction_test.go:149: invalid nonce ordering: tx #0 (A=64ba7325 N=8) < tx #13 (A=64ba7325 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #1 (A=df532874 N=6) < tx #6 (A=df532874 N=5)
	transaction_test.go:149: invalid nonce ordering: tx #1 (A=df532874 N=6) < tx #12 (A=df532874 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #1 (A=df532874 N=6) < tx #16 (A=df532874 N=3)
	transaction_test.go:149: invalid nonce ordering: tx #2 (A=df532874 N=7) < tx #6 (A=df532874 N=5)
	transaction_test.go:149: invalid nonce ordering: tx #2 (A=df532874 N=7) < tx #12 (A=df532874 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #2 (A=df532874 N=7) < tx #16 (A=df532874 N=3)
	transaction_test.go:149: invalid nonce ordering: tx #3 (A=64ba7325 N=6) < tx #5 (A=64ba7325 N=5)
	transaction_test.go:149: invalid nonce ordering: tx #3 (A=64ba7325 N=6) < tx #13 (A=64ba7325 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #4 (A=64ba7325 N=7) < tx #5 (A=64ba7325 N=5)
	transaction_test.go:149: invalid nonce ordering: tx #4 (A=64ba7325 N=7) < tx #13 (A=64ba7325 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #5 (A=fcadf686 N=5) < tx #11 (A=fcadf686 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #5 (A=fcadf686 N=5) < tx #17 (A=fcadf686 N=3)
	transaction_test.go:149: invalid nonce ordering: tx #5 (A=fcadf686 N=5) < tx #19 (A=fcadf686 N=2)
	transaction_test.go:149: invalid nonce ordering: tx #5 (A=fcadf686 N=5) < tx #23 (A=fcadf686 N=1)
	transaction_test.go:149: invalid nonce ordering: tx #6 (A=64ba7325 N=5) < tx #13 (A=64ba7325 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #7 (A=df532874 N=5) < tx #12 (A=df532874 N=4)
	transaction_test.go:149: invalid nonce ordering: tx #7 (A=df532874 N=5) < tx #16 (A=df532874 N=3)
	transaction_test.go:149: invalid nonce ordering: tx #11 (A=0834844f N=4) < tx #18 (A=0834844f N=3)
	transaction_test.go:149: invalid nonce ordering: tx #11 (A=0834844f N=4) < tx #20 (A=0834844f N=0)
	transaction_test.go:149: invalid nonce ordering: tx #11 (A=0834844f N=4) < tx #21 (A=0834844f N=1)
	transaction_test.go:149: invalid nonce ordering: tx #11 (A=0834844f N=4) < tx #22 (A=0834844f N=2)
	transaction_test.go:173: invalid gasprice ordering: tx #15 (A=63d9f865 P=5) > tx #11 (A=0834844f P=4)
	transaction_test.go:173: invalid gasprice ordering: tx #16 (A=63d9f865 P=6) > tx #11 (A=0834844f P=4)
	transaction_test.go:149: invalid nonce ordering: tx #12 (A=fcadf686 N=4) < tx #17 (A=fcadf686 N=3)
	transaction_test.go:149: invalid nonce ordering: tx #12 (A=fcadf686 N=4) < tx #19 (A=fcadf686 N=2)
	transaction_test.go:149: invalid nonce ordering: tx #12 (A=fcadf686 N=4) < tx #23 (A=fcadf686 N=1)
	transaction_test.go:170: invalid gasprice ordering: tx #8 (A=63d9f865 P=2) < tx #12 (A=fcadf686 P=4)
	transaction_test.go:170: invalid gasprice ordering: tx #9 (A=63d9f865 P=3) < tx #12 (A=fcadf686 P=4)
	transaction_test.go:173: invalid gasprice ordering: tx #15 (A=63d9f865 P=5) > tx #12 (A=fcadf686 P=4)
	transaction_test.go:173: invalid gasprice ordering: tx #16 (A=63d9f865 P=6) > tx #12 (A=fcadf686 P=4)
	transaction_test.go:149: invalid nonce ordering: tx #13 (A=df532874 N=4) < tx #16 (A=df532874 N=3)
	transaction_test.go:170: invalid gasprice ordering: tx #8 (A=63d9f865 P=2) < tx #13 (A=df532874 P=4)
	transaction_test.go:170: invalid gasprice ordering: tx #9 (A=63d9f865 P=3) < tx #13 (A=df532874 P=4)
	transaction_test.go:173: invalid gasprice ordering: tx #15 (A=63d9f865 P=5) > tx #13 (A=df532874 P=4)
	transaction_test.go:173: invalid gasprice ordering: tx #16 (A=63d9f865 P=6) > tx #13 (A=df532874 P=4)
	transaction_test.go:170: invalid gasprice ordering: tx #8 (A=63d9f865 P=2) < tx #14 (A=64ba7325 P=4)
	transaction_test.go:170: invalid gasprice ordering: tx #9 (A=63d9f865 P=3) < tx #14 (A=64ba7325 P=4)
	transaction_test.go:170: invalid gasprice ordering: tx #11 (A=0834844f P=4) < tx #15 (A=63d9f865 P=5)
	transaction_test.go:170: invalid gasprice ordering: tx #12 (A=fcadf686 P=4) < tx #15 (A=63d9f865 P=5)
	transaction_test.go:170: invalid gasprice ordering: tx #13 (A=df532874 P=4) < tx #15 (A=63d9f865 P=5)
	transaction_test.go:170: invalid gasprice ordering: tx #14 (A=64ba7325 P=4) < tx #15 (A=63d9f865 P=5)
	transaction_test.go:149: invalid nonce ordering: tx #18 (A=fcadf686 N=3) < tx #19 (A=fcadf686 N=2)
	transaction_test.go:149: invalid nonce ordering: tx #18 (A=fcadf686 N=3) < tx #23 (A=fcadf686 N=1)
	transaction_test.go:149: invalid nonce ordering: tx #19 (A=0834844f N=3) < tx #20 (A=0834844f N=0)
	transaction_test.go:149: invalid nonce ordering: tx #19 (A=0834844f N=3) < tx #21 (A=0834844f N=1)
	transaction_test.go:149: invalid nonce ordering: tx #19 (A=0834844f N=3) < tx #22 (A=0834844f N=2)
	transaction_test.go:149: invalid nonce ordering: tx #20 (A=fcadf686 N=2) < tx #23 (A=fcadf686 N=1)
	transaction_test.go:170: invalid gasprice ordering: tx #21 (A=0834844f P=0) < tx #24 (A=fcadf686 P=1)
FAIL
```

@fjl Your proposal fails with:

```
	transaction_test.go:151: invalid nonce ordering: tx #0 (A=8bc41693 N=8) < tx #0 (A=8bc41693 N=7)
	transaction_test.go:151: invalid nonce ordering: tx #0 (A=8bc41693 N=8) < tx #4 (A=8bc41693 N=6)
	transaction_test.go:151: invalid nonce ordering: tx #0 (A=8bc41693 N=8) < tx #5 (A=8bc41693 N=5)
	transaction_test.go:151: invalid nonce ordering: tx #0 (A=8bc41693 N=8) < tx #13 (A=8bc41693 N=4)
	transaction_test.go:151: invalid nonce ordering: tx #1 (A=8bc41693 N=7) < tx #4 (A=8bc41693 N=6)
	transaction_test.go:151: invalid nonce ordering: tx #1 (A=8bc41693 N=7) < tx #5 (A=8bc41693 N=5)
	transaction_test.go:151: invalid nonce ordering: tx #1 (A=8bc41693 N=7) < tx #13 (A=8bc41693 N=4)
	transaction_test.go:151: invalid nonce ordering: tx #2 (A=c078742b N=7) < tx #3 (A=c078742b N=6)
	transaction_test.go:151: invalid nonce ordering: tx #2 (A=c078742b N=7) < tx #7 (A=c078742b N=5)
	transaction_test.go:151: invalid nonce ordering: tx #2 (A=c078742b N=7) < tx #12 (A=c078742b N=4)
	transaction_test.go:151: invalid nonce ordering: tx #2 (A=c078742b N=7) < tx #14 (A=c078742b N=3)
	transaction_test.go:151: invalid nonce ordering: tx #3 (A=3e411f6c N=6) < tx #8 (A=3e411f6c N=5)
	transaction_test.go:151: invalid nonce ordering: tx #3 (A=3e411f6c N=6) < tx #9 (A=3e411f6c N=4)
	transaction_test.go:151: invalid nonce ordering: tx #3 (A=3e411f6c N=6) < tx #17 (A=3e411f6c N=3)
	transaction_test.go:151: invalid nonce ordering: tx #3 (A=3e411f6c N=6) < tx #18 (A=3e411f6c N=2)
	transaction_test.go:151: invalid nonce ordering: tx #4 (A=c078742b N=6) < tx #7 (A=c078742b N=5)
	transaction_test.go:151: invalid nonce ordering: tx #4 (A=c078742b N=6) < tx #12 (A=c078742b N=4)
	transaction_test.go:151: invalid nonce ordering: tx #4 (A=c078742b N=6) < tx #14 (A=c078742b N=3)
	transaction_test.go:151: invalid nonce ordering: tx #5 (A=8bc41693 N=6) < tx #5 (A=8bc41693 N=5)
	transaction_test.go:151: invalid nonce ordering: tx #5 (A=8bc41693 N=6) < tx #13 (A=8bc41693 N=4)
	transaction_test.go:151: invalid nonce ordering: tx #6 (A=8bc41693 N=5) < tx #13 (A=8bc41693 N=4)
	transaction_test.go:151: invalid nonce ordering: tx #7 (A=3008ac02 N=5) < tx #11 (A=3008ac02 N=4)
	transaction_test.go:151: invalid nonce ordering: tx #7 (A=3008ac02 N=5) < tx #15 (A=3008ac02 N=3)
	transaction_test.go:151: invalid nonce ordering: tx #7 (A=3008ac02 N=5) < tx #19 (A=3008ac02 N=2)
	transaction_test.go:151: invalid nonce ordering: tx #7 (A=3008ac02 N=5) < tx #22 (A=3008ac02 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #8 (A=c078742b N=5) < tx #12 (A=c078742b N=4)
	transaction_test.go:151: invalid nonce ordering: tx #8 (A=c078742b N=5) < tx #14 (A=c078742b N=3)
	transaction_test.go:151: invalid nonce ordering: tx #9 (A=3e411f6c N=5) < tx #9 (A=3e411f6c N=4)
	transaction_test.go:151: invalid nonce ordering: tx #9 (A=3e411f6c N=5) < tx #17 (A=3e411f6c N=3)
	transaction_test.go:151: invalid nonce ordering: tx #9 (A=3e411f6c N=5) < tx #18 (A=3e411f6c N=2)
	transaction_test.go:151: invalid nonce ordering: tx #10 (A=3e411f6c N=4) < tx #17 (A=3e411f6c N=3)
	transaction_test.go:151: invalid nonce ordering: tx #10 (A=3e411f6c N=4) < tx #18 (A=3e411f6c N=2)
	transaction_test.go:151: invalid nonce ordering: tx #11 (A=ce0e20a6 N=4) < tx #16 (A=ce0e20a6 N=3)
	transaction_test.go:151: invalid nonce ordering: tx #11 (A=ce0e20a6 N=4) < tx #20 (A=ce0e20a6 N=2)
	transaction_test.go:151: invalid nonce ordering: tx #11 (A=ce0e20a6 N=4) < tx #21 (A=ce0e20a6 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #11 (A=ce0e20a6 N=4) < tx #23 (A=ce0e20a6 N=0)
	transaction_test.go:151: invalid nonce ordering: tx #12 (A=3008ac02 N=4) < tx #15 (A=3008ac02 N=3)
	transaction_test.go:151: invalid nonce ordering: tx #12 (A=3008ac02 N=4) < tx #19 (A=3008ac02 N=2)
	transaction_test.go:151: invalid nonce ordering: tx #12 (A=3008ac02 N=4) < tx #22 (A=3008ac02 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #13 (A=c078742b N=4) < tx #14 (A=c078742b N=3)
	transaction_test.go:151: invalid nonce ordering: tx #16 (A=3008ac02 N=3) < tx #19 (A=3008ac02 N=2)
	transaction_test.go:151: invalid nonce ordering: tx #16 (A=3008ac02 N=3) < tx #22 (A=3008ac02 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #17 (A=ce0e20a6 N=3) < tx #20 (A=ce0e20a6 N=2)
	transaction_test.go:151: invalid nonce ordering: tx #17 (A=ce0e20a6 N=3) < tx #21 (A=ce0e20a6 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #17 (A=ce0e20a6 N=3) < tx #23 (A=ce0e20a6 N=0)
	transaction_test.go:151: invalid nonce ordering: tx #18 (A=3e411f6c N=3) < tx #18 (A=3e411f6c N=2)
	transaction_test.go:151: invalid nonce ordering: tx #20 (A=3008ac02 N=2) < tx #22 (A=3008ac02 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #21 (A=ce0e20a6 N=2) < tx #21 (A=ce0e20a6 N=1)
	transaction_test.go:151: invalid nonce ordering: tx #21 (A=ce0e20a6 N=2) < tx #23 (A=ce0e20a6 N=0)
	transaction_test.go:151: invalid nonce ordering: tx #22 (A=ce0e20a6 N=1) < tx #23 (A=ce0e20a6 N=0)
```

Inserting the sort order would fail with:

```
--- FAIL: TestTransactionPriceNonceSort (0.01s)
	transaction_test.go:175: invalid gasprice ordering: tx #3 (A=197d46b4 P=2) > tx #1 (A=b17b67ef P=1)
	transaction_test.go:175: invalid gasprice ordering: tx #4 (A=73cb4a16 P=2) > tx #1 (A=b17b67ef P=1)
	transaction_test.go:175: invalid gasprice ordering: tx #3 (A=197d46b4 P=2) > tx #2 (A=73cb4a16 P=1)
	transaction_test.go:175: invalid gasprice ordering: tx #6 (A=b17b67ef P=3) > tx #3 (A=197d46b4 P=2)
	transaction_test.go:175: invalid gasprice ordering: tx #7 (A=1e6e4f2f P=3) > tx #3 (A=197d46b4 P=2)
	transaction_test.go:175: invalid gasprice ordering: tx #8 (A=73cb4a16 P=3) > tx #3 (A=197d46b4 P=2)
	transaction_test.go:175: invalid gasprice ordering: tx #6 (A=b17b67ef P=3) > tx #4 (A=73cb4a16 P=2)
	transaction_test.go:175: invalid gasprice ordering: tx #7 (A=1e6e4f2f P=3) > tx #4 (A=73cb4a16 P=2)
	transaction_test.go:172: invalid gasprice ordering: tx #2 (A=73cb4a16 P=1) < tx #5 (A=b17b67ef P=2)
	transaction_test.go:175: invalid gasprice ordering: tx #10 (A=197d46b4 P=4) > tx #6 (A=b17b67ef P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #10 (A=197d46b4 P=4) > tx #7 (A=1e6e4f2f P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #11 (A=b17b67ef P=4) > tx #7 (A=1e6e4f2f P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #12 (A=73cb4a16 P=4) > tx #7 (A=1e6e4f2f P=3)
	transaction_test.go:172: invalid gasprice ordering: tx #5 (A=b17b67ef P=2) < tx #8 (A=73cb4a16 P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #10 (A=197d46b4 P=4) > tx #8 (A=73cb4a16 P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #11 (A=b17b67ef P=4) > tx #8 (A=73cb4a16 P=3)
	transaction_test.go:172: invalid gasprice ordering: tx #4 (A=73cb4a16 P=2) < tx #9 (A=197d46b4 P=3)
	transaction_test.go:172: invalid gasprice ordering: tx #5 (A=b17b67ef P=2) < tx #9 (A=197d46b4 P=3)
	transaction_test.go:175: invalid gasprice ordering: tx #15 (A=73cb4a16 P=5) > tx #10 (A=197d46b4 P=4)
	transaction_test.go:175: invalid gasprice ordering: tx #16 (A=1e6e4f2f P=5) > tx #10 (A=197d46b4 P=4)
	transaction_test.go:175: invalid gasprice ordering: tx #17 (A=cca60613 P=5) > tx #10 (A=197d46b4 P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #7 (A=1e6e4f2f P=3) < tx #11 (A=b17b67ef P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #8 (A=73cb4a16 P=3) < tx #11 (A=b17b67ef P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #9 (A=197d46b4 P=3) < tx #11 (A=b17b67ef P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #9 (A=197d46b4 P=3) < tx #12 (A=73cb4a16 P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #8 (A=73cb4a16 P=3) < tx #13 (A=1e6e4f2f P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #9 (A=197d46b4 P=3) < tx #13 (A=1e6e4f2f P=4)
	transaction_test.go:175: invalid gasprice ordering: tx #15 (A=73cb4a16 P=5) > tx #13 (A=1e6e4f2f P=4)
	transaction_test.go:175: invalid gasprice ordering: tx #15 (A=73cb4a16 P=5) > tx #14 (A=cca60613 P=4)
	transaction_test.go:175: invalid gasprice ordering: tx #16 (A=1e6e4f2f P=5) > tx #14 (A=cca60613 P=4)
	transaction_test.go:172: invalid gasprice ordering: tx #13 (A=1e6e4f2f P=4) < tx #15 (A=73cb4a16 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #14 (A=cca60613 P=4) < tx #15 (A=73cb4a16 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #14 (A=cca60613 P=4) < tx #16 (A=1e6e4f2f P=5)
	transaction_test.go:175: invalid gasprice ordering: tx #19 (A=197d46b4 P=6) > tx #16 (A=1e6e4f2f P=5)
	transaction_test.go:175: invalid gasprice ordering: tx #19 (A=197d46b4 P=6) > tx #17 (A=cca60613 P=5)
	transaction_test.go:175: invalid gasprice ordering: tx #20 (A=1e6e4f2f P=6) > tx #17 (A=cca60613 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #11 (A=b17b67ef P=4) < tx #18 (A=197d46b4 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #12 (A=73cb4a16 P=4) < tx #18 (A=197d46b4 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #13 (A=1e6e4f2f P=4) < tx #18 (A=197d46b4 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #14 (A=cca60613 P=4) < tx #18 (A=197d46b4 P=5)
	transaction_test.go:172: invalid gasprice ordering: tx #17 (A=cca60613 P=5) < tx #20 (A=1e6e4f2f P=6)
	transaction_test.go:172: invalid gasprice ordering: tx #18 (A=197d46b4 P=5) < tx #20 (A=1e6e4f2f P=6)
	transaction_test.go:175: invalid gasprice ordering: tx #22 (A=cca60613 P=7) > tx #20 (A=1e6e4f2f P=6)
	transaction_test.go:172: invalid gasprice ordering: tx #18 (A=197d46b4 P=5) < tx #21 (A=cca60613 P=6)
	transaction_test.go:172: invalid gasprice ordering: tx #21 (A=cca60613 P=6) < tx #23 (A=1e6e4f2f P=7)
	transaction_test.go:172: invalid gasprice ordering: tx #23 (A=1e6e4f2f P=7) < tx #24 (A=cca60613 P=8)
```

I'll try to write the heap-merge based algo and see how that works.

--- a/core/transaction_pool.go
+++ b/core/transaction_pool.go
@@ -376,7 +376,7 @@ func (self *TxPool) GetQueuedTransactions() types.Transactions {
 			ret = append(ret, tx)
 		}
 	}
-	sort.Sort(types.TxByNonce{ret})
+	sort.Sort(types.TxByNonce(ret))
 	return ret
 }
 

--- a/core/types/transaction.go
+++ b/core/types/transaction.go
@@ -17,11 +17,13 @@
 package types
 
 import (
+	"container/heap"
 	"crypto/ecdsa"
 	"errors"
 	"fmt"
 	"io"
 	"math/big"
+	"sort"
 	"sync/atomic"
 
 	"github.com/ethereum/go-ethereum/common"
@@ -302,27 +304,78 @@ func TxDifference(a, b Transactions) (keep Transactions) {
 	return keep
 }
 
-type TxByNonce struct{ Transactions }
+// TxByNonce implements the sort interface to allow sorting a list of transactions
+// by their nonces. This is usually only useful for sorting transactions from a
+// single account, otherwise a nonce comparison doesn't make much sense.
+type TxByNonce Transactions
 
-func (s TxByNonce) Less(i, j int) bool {
-	return s.Transactions[i].data.AccountNonce < s.Transactions[j].data.AccountNonce
+func (s TxByNonce) Len() int           { return len(s) }
+func (s TxByNonce) Less(i, j int) bool { return s[i].data.AccountNonce < s[j].data.AccountNonce }
+func (s TxByNonce) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// TxByPrice implements both the sort and the heap interface, making it useful
+// for all at once sorting as well as individually adding and removing elements.
+type TxByPrice Transactions
+
+func (s TxByPrice) Len() int           { return len(s) }
+func (s TxByPrice) Less(i, j int) bool { return s[i].data.Price.Cmp(s[j].data.Price) > 0 }
+func (s TxByPrice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+func (s *TxByPrice) Push(x interface{}) {
+	*s = append(*s, x.(*Transaction))
 }
 
-type TxByPrice struct{ Transactions }
-
-func (s TxByPrice) Less(i, j int) bool {
-	return s.Transactions[i].data.Price.Cmp(s.Transactions[j].data.Price) > 0
+func (s *TxByPrice) Pop() interface{} {
+	old := *s
+	n := len(old)
+	x := old[n-1]
+	*s = old[0 : n-1]
+	return x
 }
 
-type TxByPriceAndNonce struct{ Transactions }
-
-func (s TxByPriceAndNonce) Less(i, j int) bool {
-	// we can ignore the error here. Sorting shouldn't care about validness
-	ifrom, _ := s.Transactions[i].From()
-	jfrom, _ := s.Transactions[j].From()
-	// favour nonce if they are from the same recipient
-	if ifrom == jfrom {
-		return s.Transactions[i].data.AccountNonce < s.Transactions[j].data.AccountNonce
+// SortByPriceAndNonce sorts the transactions by price in such a way that the
+// nonce orderings within a single account are maintained.
+//
+// Note, this is not as trivial as it seems from the first look as there are three
+// different criteria that need to be taken into account (price, nonce, account
+// match), which cannot be done with any plain sorting method, as certain items
+// cannot be compared without context.
+//
+// This method first sorts the separates the list of transactions into individual
+// sender accounts and sorts them by nonce. After the account nonce ordering is
+// satisfied, the results are merged back together by price, always comparing only
+// the head transaction from each account. This is done via a heap to keep it fast.
+func SortByPriceAndNonce(txs []*Transaction) {
+	// Separate the transactions by account and sort by nonce
+	byNonce := make(map[common.Address][]*Transaction)
+	for _, tx := range txs {
+		acc, _ := tx.From() // we only sort valid txs so this cannot fail
+		byNonce[acc] = append(byNonce[acc], tx)
 	}
-	return s.Transactions[i].data.Price.Cmp(s.Transactions[j].data.Price) > 0
+	for _, accTxs := range byNonce {
+		sort.Sort(TxByNonce(accTxs))
+	}
+	// Initialize a price based heap with the head transactions
+	byPrice := make(TxByPrice, 0, len(byNonce))
+	for acc, accTxs := range byNonce {
+		byPrice = append(byPrice, accTxs[0])
+		byNonce[acc] = accTxs[1:]
+	}
+	heap.Init(&byPrice)
+
+	// Merge by replacing the best with the next from the same account
+	txs = txs[:0]
+	for len(byPrice) > 0 {
+		// Retrieve the next best transaction by price
+		best := heap.Pop(&byPrice).(*Transaction)
+
+		// Push in its place the next transaction from the same account
+		acc, _ := best.From() // we only sort valid txs so this cannot fail
+		if accTxs, ok := byNonce[acc]; ok && len(accTxs) > 0 {
+			heap.Push(&byPrice, accTxs[0])
+			byNonce[acc] = accTxs[1:]
+		}
+		// Accumulate the best priced transaction
+		txs = append(txs, best)
+	}
 }

--- a/core/types/transaction_test.go
+++ b/core/types/transaction_test.go
@@ -117,3 +117,60 @@ func TestRecipientNormal(t *testing.T) {
 		t.Error("derived address doesn't match")
 	}
 }
+
+// Tests that transactions can be correctly sorted according to their price in
+// decreasing order, but at the same time with increasing nonces when issued by
+// the same account.
+func TestTransactionPriceNonceSort(t *testing.T) {
+	// Generate a batch of accounts to start with
+	keys := make([]*ecdsa.PrivateKey, 25)
+	for i := 0; i < len(keys); i++ {
+		keys[i], _ = crypto.GenerateKey()
+	}
+	// Generate a batch of transactions with overlapping values, but shifted nonces
+	txs := []*Transaction{}
+	for start, key := range keys {
+		for i := 0; i < 25; i++ {
+			tx, _ := NewTransaction(uint64(start+i), common.Address{}, big.NewInt(100), big.NewInt(100), big.NewInt(int64(start+i)), nil).SignECDSA(key)
+			txs = append(txs, tx)
+		}
+	}
+	// Sort the transactions and cross check the nonce ordering
+	SortByPriceAndNonce(txs)
+	for i, txi := range txs {
+		fromi, _ := txi.From()
+
+		// Make sure the nonce order is valid
+		for j, txj := range txs[i+1:] {
+			fromj, _ := txj.From()
+
+			if fromi == fromj && txi.Nonce() > txj.Nonce() {
+				t.Errorf("invalid nonce ordering: tx #%d (A=%x N=%v) < tx #%d (A=%x N=%v)", i, fromi[:4], txi.Nonce(), i+j, fromj[:4], txj.Nonce())
+			}
+		}
+		// Find the previous and next nonce of this account
+		prev, next := i-1, i+1
+		for j := i - 1; j >= 0; j-- {
+			if fromj, _ := txs[j].From(); fromi == fromj {
+				prev = j
+				break
+			}
+		}
+		for j := i + 1; j < len(txs); j++ {
+			if fromj, _ := txs[j].From(); fromi == fromj {
+				next = j
+				break
+			}
+		}
+		// Make sure that in between the neighbor nonces, the transaction is correctly positioned price wise
+		for j := prev + 1; j < next; j++ {
+			fromj, _ := txs[j].From()
+			if j < i && txs[j].GasPrice().Cmp(txi.GasPrice()) < 0 {
+				t.Errorf("invalid gasprice ordering: tx #%d (A=%x P=%v) < tx #%d (A=%x P=%v)", j, fromj[:4], txs[j].GasPrice(), i, fromi[:4], txi.GasPrice())
+			}
+			if j > i && txs[j].GasPrice().Cmp(txi.GasPrice()) > 0 {
+				t.Errorf("invalid gasprice ordering: tx #%d (A=%x P=%v) > tx #%d (A=%x P=%v)", j, fromj[:4], txs[j].GasPrice(), i, fromi[:4], txi.GasPrice())
+			}
+		}
+	}
+}

--- a/miner/worker.go
+++ b/miner/worker.go
@@ -19,7 +19,6 @@ package miner
 import (
 	"fmt"
 	"math/big"
-	"sort"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -496,12 +495,12 @@ func (self *worker) commitNewWork() {
 
 	/* //approach 1
 	transactions := self.eth.TxPool().GetTransactions()
-	sort.Sort(types.TxByNonce{transactions})
+	sort.Sort(types.TxByNonce(transactions))
 	*/
 
 	//approach 2
 	transactions := self.eth.TxPool().GetTransactions()
-	sort.Sort(types.TxByPriceAndNonce{transactions})
+	types.SortByPriceAndNonce(transactions)
 
 	/* // approach 3
 	// commit transactions for this run.
@@ -525,8 +524,8 @@ func (self *worker) commitNewWork() {
 			multiTxOwner = append(multiTxOwner, txs...)
 		}
 	}
-	sort.Sort(types.TxByPrice{singleTxOwner})
-	sort.Sort(types.TxByNonce{multiTxOwner})
+	sort.Sort(types.TxByPrice(singleTxOwner))
+	sort.Sort(types.TxByNonce(multiTxOwner))
 	transactions := append(singleTxOwner, multiTxOwner...)
 	*/
 

